File: tests/unit/src/main/scala/tests/FileLayout.scala
package tests

import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.nio.file.FileVisitResult
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.SimpleFileVisitor
import java.nio.file.StandardOpenOption
import java.nio.file.attribute.BasicFileAttributes

import scala.collection.mutable

/** Converts between a text "file layout" description and real files on disk. */
object FileLayout {

  def asString(
      root: Path,
      includePath: Path => Boolean = _ => true,
      charset: Charset = StandardCharsets.UTF_8
  ): String = {
    if (!Files.isDirectory(root))
      return ""
    import scala.jdk.CollectionConverters._
    val out = new StringBuilder()
    val buf = mutable.ArrayBuffer.empty[Path]
    Files.walkFileTree(
      root,
      new SimpleFileVisitor[Path] {
        override def preVisitDirectory(
            dir: Path,
            attrs: BasicFileAttributes
        ): FileVisitResult =
          if (!includePath(dir))
            FileVisitResult.SKIP_SUBTREE
          else
            FileVisitResult.CONTINUE

        override def visitFile(
            file: Path,
            attrs: BasicFileAttributes
        ): FileVisitResult = {
          if (includePath(file))
            buf += file
          FileVisitResult.CONTINUE
        }
      }
    )
    buf
      .sorted
      .foreach { file =>
        val relpath = root.relativize(file).iterator().asScala.mkString("/")
        out.append("/").append(relpath)
        if (Files.isSymbolicLink(file)) {
          out.append(" -> ").append(Files.readSymbolicLink(file))
        } else {
          val bytes = Files.readAllBytes(file)
          out.append("\n").append(new String(bytes, charset))
        }
        out.append("\n")
      }
    out.toString()
  }

  def mapFromString(layout: String): Map[String, String] = {
    if (layout.trim.isEmpty)
      return Map.empty
    val normalized = layout.replace("\r\n", "\n")
    normalized
      .split("(?=\n/)")
      .map { row =>
        row.stripPrefix("\n").split("\n", 2).toList match {
          case path :: contents :: Nil =>
            val withEndOfFileLine =
              if (contents.endsWith("\n"))
                contents
              else
                contents + "\n"
            path.stripPrefix("/") -> withEndOfFileLine
          case other =>
            throw new IllegalArgumentException(
              s"Unable to split argument into path/contents! \n$other"
            )
        }
      }
      .toMap
  }

  def fromString(
      layout: String,
      root: Path = Files.createTempDirectory("scip-java"),
      charset: Charset = StandardCharsets.UTF_8
  ): Path = {
    if (layout.trim.isEmpty)
      return root
    mapFromString(layout).foreach { case (path, contents) =>
      val file = path.split("/").foldLeft(root)(_ resolve _)
      val parent = file.getParent
      // Don't try to mkdir under a symlinked parent.
      if (!Files.exists(parent))
        Files.createDirectories(parent)
      Files.deleteIfExists(file)
      Files.write(
        file,
        contents.getBytes(charset),
        StandardOpenOption.WRITE,
        StandardOpenOption.CREATE
      )
    }
    root
  }
}

File: tests/buildtools-kotlin/src/test/kotlin/tests/FileLayout.kt
package tests

import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption

/** Converts a text "file layout" description into real files on disk. */
object FileLayout {

    fun mapFromString(layout: String): Map<String, String> {
        if (layout.trim().isEmpty()) return emptyMap()
        val normalized = layout.replace("\r\n", "\n")
        return splitBeforeEntries(normalized).associate { row ->
            val parts = row.removePrefix("\n").split("\n", limit = 2)
            require(parts.size == 2) {
                "Unable to split argument into path/contents!\n$row"
            }
            val (path, contents) = parts
            val withEndOfFileLine =
                if (contents.endsWith("\n")) contents else contents + "\n"
            path.removePrefix("/") to withEndOfFileLine
        }
    }

    /**
     * Splits before every `\n/` boundary, matching the Scala harness'
     * `split("(?=\n/)")`. A boundary at index 0 is kept in the first row (it is
     * stripped later by `removePrefix("\n")`) rather than producing an empty
     * leading row, which is where Kotlin's `Regex.split` would differ.
     */
    private fun splitBeforeEntries(s: String): List<String> {
        val rows = mutableListOf<String>()
        var start = 0
        var i = 1
        while (i < s.length) {
            if (s[i] == '\n' && i + 1 < s.length && s[i + 1] == '/') {
                rows.add(s.substring(start, i))
                start = i
            }
            i++
        }
        rows.add(s.substring(start))
        return rows
    }

    fun fromString(
        layout: String,
        root: Path,
        charset: Charset = StandardCharsets.UTF_8,
    ): Path {
        if (layout.trim().isEmpty()) return root
        for ((path, contents) in mapFromString(layout)) {
            val file = path.split("/").fold(root) { acc, part -> acc.resolve(part) }
            val parent = file.parent
            if (!Files.exists(parent)) Files.createDirectories(parent)
            Files.deleteIfExists(file)
            Files.write(
                file,
                contents.toByteArray(charset),
                StandardOpenOption.WRITE,
                StandardOpenOption.CREATE,
            )
        }
        return root
    }
}
